Certificate Signing Request Issuer CA

JoshVanL opened the pull request:
**What this PR does / why we need it**:
This PR adds an _optional_ CertificateSigningRequest CA issuer. The controller is only enabled if it is explicitly done so via the `--feature-gates` flag (i.e. `--feature-gates=ExperimentalCertificateSigningRequestControllers=true`)

This begins the work to integrating kubernetes CertificateSigningRequests as part of cert-manager. (ref #3646).

Users who request a certificate referencing a namespaced issuer, must have permissions to do so in the following form:

```
group: cert-manager.io
resource: signers
verb: reference
namespace: <referenced signer namespace>
name: <either the name of the signer or '*' for all signer names in that namespace>
```

A SubjectAccessReview is performed during processing to check whether the user has sufficient permissions to reference the Issuer encoded into the signer name.

**Special notes for your reviewer**:

This work is based on the design in #3646 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Adds an option for a Kubernetes CertificateSigningRequest controller to implement the CA Issuer.
```

/cc @munnerz

JoshVanL commented:
flake
/retest

irbekrm commented:
/test pull-cert-manager-e2e-v1-21

SgtCoDFish commented on pkg/controller/certificatesigningrequests/util/signername.go:
```diff
@@ -0,0 +1,63 @@
+	}
+
+	signerTypeSplit := strings.SplitN(split[0], ".", 2)
+	signerNameSplit := strings.Split(split[1], ".")
```
suggestion(important): If I understand correctly, I think this function is important as we're defining how cert-manger issuers can be referenced as signers, and I think we're saying that to specify an Issuer, we want the format `foo.bar/namespace.issuerName`.

I think this function would return false with the following (valid) issuer:

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: mynamespace
spec:
  finalizers:
  - kubernetes
apiVersion: v1

---

apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned.issuer
  namespace: mynamespace
spec:
  selfSigned: {}
```

I think this would be specified as `foo.bar/mynamespace.selfsigned.issuer`;  `.` is a valid character in an issuer name (but not in the name for a namespace).

Should the switch statement here instead be something like:

```go
if len(signerNameSplit) == 1 {
	return SignerIssuerRef{
    		Namespace: "",
    		Name:      signerNameSplit[0],
    		Type:      signerTypeSplit[0],
    		Group:     signerTypeSplit[1],
	}, true
}

return SignerIssuerRef{
   	Namespace: signerNameSplit[0],
    	Name:      strings.Join(signerNameSplit[1:], "."),
   	Type:      signerTypeSplit[0],
   	Group:     signerTypeSplit[1],
}, true
```

SgtCoDFish replied:
this was by far the most important thing in the review, so good to see it's addressed.

SgtCoDFish commented on pkg/controller/certificatesigningrequests/util/signername_test.go:
```diff
@@ -0,0 +1,116 @@
+		expSignerIssuerRef SignerIssuerRef
+		expOK              bool
+	}{
+		"an empty name should return false": {
```
suggestion: does this need test cases for issuers + clusterissuers with a "." in their names?

JoshVanL replied:
Not sure I follow

SgtCoDFish replied:
https://github.com/jetstack/cert-manager/pull/4064#discussion_r640683648 <- test cases for this.

SgtCoDFish replied:
I see you added test cases for this so maybe you understood subconsciously :grin:

SgtCoDFish commented on pkg/util/pki/kube.go:
```diff
@@ -0,0 +1,75 @@
+		}
+		duration = dur
+	}
+	fmt.Println(duration)
```
question: was printing the duration intentional here?

SgtCoDFish commented on pkg/util/pki/kube.go:
```diff
@@ -0,0 +1,75 @@
+
+// GenerateTemplate will create a x509.Certificate for the given
+// CertificateSigningRequest resource
+func GenerateTemplateFromCertificateSigningRequest(csr *certificatesv1.CertificateSigningRequest) (*x509.Certificate, error) {
```
suggestion: I think this function could use some unit tests

SgtCoDFish commented on pkg/util/pki/kube.go:
```diff
@@ -0,0 +1,75 @@
+	experimentalapi "github.com/jetstack/cert-manager/pkg/apis/experimental/v1alpha1"
+)
+
+// GenerateTemplate will create a x509.Certificate for the given
+// CertificateSigningRequest resource
```
nitpick: doc comment nitpick

```suggestion
// GenerateTemplateFromCertificateSigningRequest will create an *x509.Certificate from the given
// CertificateSigningRequest resource
```

SgtCoDFish commented on pkg/util/pki/kube.go:
```diff
@@ -0,0 +1,75 @@
+	return GenerateTemplateFromCSRPEMWithUsages(csr.Spec.Request, duration, isCA, ku, eku)
+}
+
+func BuildKeyUsagesKube(usages []certificatesv1.KeyUsage) (ku x509.KeyUsage, eku []x509.ExtKeyUsage, err error) {
```
suggestion: I think this function would really benefit from some unit tests.

JoshVanL replied:
Done :)

SgtCoDFish commented on pkg/util/pki/kube.go:
```diff
@@ -0,0 +1,75 @@
+func BuildKeyUsagesKube(usages []certificatesv1.KeyUsage) (ku x509.KeyUsage, eku []x509.ExtKeyUsage, err error) {
+	var unk []certificatesv1.KeyUsage
+	if len(usages) == 0 {
+		usages = []certificatesv1.KeyUsage{certificatesv1.UsageDigitalSignature, certificatesv1.UsageKeyEncipherment}
```
suggestion: should these default values come from https://github.com/jetstack/cert-manager/blob/a80198c03d108e4ceb7e5aceb0f2fd78be5db39d/pkg/apis/certmanager/v1/types.go#L203?

(I recognise that will require converting a `certmangerv1.KeyUsage` to a `certificatesv1.KeyUsage`, but I think it'd be good to re-use that function)

JoshVanL replied:
Not sure I understand the intention with reusing this function. We are using a separate API group and are different types.

SgtCoDFish replied:
"Default cert-manager key usages for certificates" is a useful concept and if we're going to have it as a concept we might as well have one place to change it rather than multiple places where we can forget one of them. I suppose we'd have, say, in pseudocode:

```go
func DefaultKeyUsages() []x509.KeyUsage
```

in pkg/internal/ somewhere. I don't think it's required for this PR to be merged.

JoshVanL replied:
Let's leave this for now. We can always add the Default to the API at a later date.

SgtCoDFish reviewed (changes requested):
a few comments from me; a lot of these are small but there are a couple of bigger ones.

as this is quite a big PR[1], I can't be fully confident that I've been over everything - a second set of eyes would be good here I think.

[1] not a criticism; it's hard to imagine how this could have been cut down much.

JoshVanL commented on deploy/charts/cert-manager/templates/rbac.yaml:
```diff
@@ -504,4 +504,57 @@ subjects:
 
+---
+
+# Permission to:
+# - Update and sign CertificatSigningeRequests referencing cert-manager.io Issuers and ClusterIssuers
+# - Perform SubjectAccessReviews to test whether users are able to reference Namespaced Issuers
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
```
It would. Is that really a concern?

JoshVanL replied:
SubjectAccessReviews are fairly benign for cert-manager to have, and it is only allowed to sign CSRs which explicitly reference cert-manager signers.

SgtCoDFish replied:
It's the kind of thing that makes security teams twitchy because it breaks the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege). Plus, once claimed, it's hard to roll back additional privileges in the future.

I'm not overly concerned by these specific permissions, though - I can just see why a security engineer might be surprised that cert-manager can update CSRs that it never uses.

JoshVanL commented on pkg/controller/certificatesigningrequests/ca/ca.go:
```diff
@@ -0,0 +1,157 @@
+type templateGenerator func(*certificatesv1.CertificateSigningRequest) (*x509.Certificate, error)
+type signingFn func([]*x509.Certificate, crypto.Signer, *x509.Certificate) (pki.PEMBundle, error)
+
+type CA struct {
```
Thought it was clear since it was under the `/controller` path, but will add some comments.

JoshVanL commented on pkg/controller/certificatesigningrequests/ca/ca_test.go:
```diff
@@ -0,0 +1,756 @@
+		t.Fatal(err)
+	}
+
+	testpk, err := pki.GenerateECPrivateKey(256)
```
Not sure what you mean

SgtCoDFish replied:
other variables are named according to the go standard, e.g. they'd be "testPK" while this one doesn't follow those standards. it doesn't really matter, I think it was copy-pasted.

irbekrm reviewed (changes requested):
Looks good to me- I have read through the code and ran the tests.
I've added a couple nits, but generally happy to approve.

JoshVanL commented on pkg/controller/certificatesigningrequests/controller_test.go:
```diff
@@ -0,0 +1,674 @@
+			existingIssuer: nil,
+			wantEvent:      "Warning IssuerNotFound Referenced Issuer hello/world not found",
+		},
+		"if CertificateSigningRequest references an Issuer that does not yet have a type, should fire an event it doesn't have a type": {
```
Not really. We do validation at admission for an Issuer, but good to check anyway.

JoshVanL commented on pkg/controller/certificatesigningrequests/sync.go:
```diff
@@ -0,0 +1,197 @@
+		dbg.Info("certificate signing request has failed so skipping processing")
+		return nil
+	}
+	if !util.CertificateSigningRequestIsApproved(csr) {
```
Yep, this is enforced by the API server.

JoshVanL commented:
Thanks @SgtCoDFish @irbekrm 

I have addressed your comments in the latest commit or responded. Good for another review and (hopefully) merge :slightly_smiling_face:

JoshVanL commented:
/assign @irbekrm @SgtCoDFish

irbekrm reviewed (approved):
Looks good to me- I could have spent more time actually trying it out, but I guess since it's an experimental controller it'd be easier to change things later.

SgtCoDFish commented on pkg/util/pki/kube_test.go:
```diff
@@ -0,0 +1,205 @@
+		},
+		"a CSR that contains invalid usages should return an error": {
+			csr: gen.CertificateSigningRequest("",
+				gen.SetCertificateSigningRequestDuration("bad-duration"),
```
suggestion: should this duration be valid?

JoshVanL replied:
Yep, thanks.

SgtCoDFish commented on pkg/apis/experimental/doc.go:
```diff
@@ -17,7 +17,7 @@ limitations under the License.
 // +groupGoName=Experimental
 
-// Package certmanager is the internal version of the API.
+// Package experimental is the internal version of the API.
```
nitpick(non-blocking): the package name is correct now, but the comment itself doesn't refer to what this package does; it's still copy-pasted.

SgtCoDFish reviewed (changes requested):
just one test looks incorrect to me

JoshVanL commented on pkg/controller/certificatesigningrequests/util/signername.go:
```diff
@@ -0,0 +1,71 @@
+		return SignerIssuerRef{}, false
+	}
+
+	switch len(signerNameSplit) {
+	case 1:
+		return SignerIssuerRef{
+			Namespace: "",
+			Name:      signerNameSplit[0],
+			Type:      signerTypeSplit[0],
+			Group:     signerTypeSplit[1],
+		}, true
+
+	default:
+		// ClusterIssuers do not have Namespaces
+		if signerTypeSplit[0] == "clusterissuers" {
+			return SignerIssuerRef{
+				Namespace: "",
+				Name:      strings.Join(signerNameSplit[0:], "."),
+				Type:      signerTypeSplit[0],
+				Group:     signerTypeSplit[1],
+			}, true
+		}
+
+		// Non Cluster Scoped issuers always have Namespaces
+		return SignerIssuerRef{
+			Namespace: signerNameSplit[0],
+			Name:      strings.Join(signerNameSplit[1:], "."),
+			Type:      signerTypeSplit[0],
+			Group:     signerTypeSplit[1],
+		}, true
+	}
```
I'm a big proponent for switch statements personally :shrug:

JoshVanL replied:
Happy to change it to an `if` statement though.

SgtCoDFish reviewed (approved):
/lgtm

SgtCoDFish commented:
/test pull-cert-manager-e2e-v1-21
